### ISSUE TITLE
Support for Argon2 KDF

### ISSUE THREAD

**Issue by johndoe31415** [openssl/openssl#4091]
I think it would be useful to have the Argon2 KDF contained within OpenSSL. It's the [winner of the PHC](https://password-hashing.net/) and the [recommendation of OWASP for new applications](https://www.owasp.org/index.php/Password_Storage_Cheat_Sheet).

There is [reference code available](https://github.com/p-h-c/phc-winner-argon2) that is CC-0 licensed. So I'm guessing this is incompatible with OpenSSL and would require a rewrite?

Since I've poked around the last few days in the KDF code, I'd like to give this a shot as well and provide a PR.

**Comment by richsalz** [comment]
Look at https://datatracker.ietf.org/doc/draft-irtf-cfrg-argon2/ for a text description, sample code (BSD license so okay to use as a starting point) and test vectors that can be used in the test suite.

**Comment by johndoe31415** [comment]
Hi @richsalz, thanks for the pointer. So I guess this would be something that is of interest for inclusion?

For efficient computation of Argon2, the ability to run concurrent code is necessary. While OpenSSL abstracts some threading functions away from their OS libraries (such as NPTL), I do not see a way to create and wait for threads. Do you have an idea on how I should approach this?

**Comment by richsalz** [comment]
Don't mistake my giving a link for general interest :)  @mattcaswell can best answer the thread questions, but if argon2 *requires* threads and will not work without them, then it's of much less interest.

**Comment by johndoe31415** [comment]
I see :-)

Well, Argon2 doen't strictly require threads in the sense that it isn't computable without them. However, it is designed such that internally it has comutation that is entirely data-independent (and, thus, perfectly parallelizable). Of course this can be sequentialized, but would essentially mean a factor of n slowdown (this is the parametrizable "lane" paramter of Argon2). And slow means useless in that context (because an attacker obviously will use a highly optimized version that does make use of the parallelizability).

I do think it's something that'd be nice in OpenSSL (especially since it's becoming popular quickly), but before I'm doing actual work, I'd then be waiting for further input on this.

**Comment by mbroz** [comment]
Just FYI: [cryptsetup](https://github.com/mbroz/cryptsetup) is one of the Argon2 users already - LUKS2 uses Argon2 as a KDF.

For now we use either system libargon2 (if available) or embedded reference implementation. If we can use OpenSSL directly, it would be much simpler (we can already use PBKDF2 from OpenSSL).

Just it should be the fully threaded version - while libargon2 code supports compilation without threads, this does not make much sense - the function has 3 costs: iterations (cpu cost), memory cost (it is memory-hard function) and parallel cost (threads). Code should support all of them.

**Comment by maqp** [comment]
Hi! First of all, thank you for your hard work and congrats on the 1.1.1 release. The [release blog post](https://www.openssl.org/blog/blog/2018/09/11/release111/) says

>The OpenSSL team will now be moving our focus to the next release which will see us developing a new FIPS module.

Does this mean that adding support for Argon2 will be postponed?

**Comment by mattcaswell** [comment]
> Does this mean that adding support for Argon2 will be postponed?

No, not necessarily. The master branch is currently open for new features. The primary focus of the dev team is the FIPS module but that doesn't mean we won't merge other stuff (in much the same way that the primary focus for the last release was TLSv1.3, but a load of other features went in too). However since this isn't a priority item it is more likely to get merged if someone contributes a PR.

**Comment by johndoe31415** [comment]
Hey @mattcaswell -- could you maybe clarify on one part: Argon2 *only* makes sense when it can utilize concurrent computation. It was my understanding that this was the major blocker to its inclusion and it's also the reason why this pretty much came to a screeching halt even before anyone had written a line of code. Has this changed?

**Comment by mattcaswell** [comment]
> Hey @mattcaswell -- could you maybe clarify on one part: Argon2 only makes sense when it can utilize concurrent computation. It was my understanding that this was the major blocker to its inclusion and it's also the reason why this pretty much came to a screeching halt even before anyone had written a line of code. Has this changed?

No, my comment above was only meant to imply that the master branch is currently open to new features (whereas it has been frozen to features for sometime until yesterday). The concurrency issue would still need to be resolved. I don't have a good answer for that one.

**Comment by maqp** [comment]
@johndoe31415, @mattcaswell The [reference implementation](https://github.com/p-h-c/phc-winner-argon2) of Argon2 supports parallelism. Is there something I'm missing here that prevents using their code?

**Comment by levitte** [comment]
Even if there is a reference implementation, it still needs work to get integrated in the openssl ecosystem. That will happen in form of a PR, regardless of who does it.

So now it only takes someone making it enough of a priority. That could be you.

**Comment by mattcaswell** [comment]
> The reference implementation of Argon2 supports parallelism. Is there something I'm missing here that prevents using their code?

The creation and destruction of threads inside the library is not something we currently do. There is no thread pool mechanism to support that and I'm not convinced we would want to include one.

**Comment by mattcaswell** [comment]
Of course the other problem with Argon2 is that AFAIK it is not standardised. The internet draft for it seems to have expired. It is a requirement for OpenSSL that all algorithms are standardised.

**Comment by levitte** [comment]
... or at least going steadily toward standardisation (see our TLSv1. 3 effort)

**Comment by mbroz** [comment]
According to my information, there is still an intent to work on Argon2 RFC (and I asked authors again).

The reference implementation supports no threads option, but for real-world implementation you will need to support threads or some equivalents. But as the first step I think it can be implemented without thread support.

**Comment by kaduk** [comment]
> Of course the other problem with Argon2 is that AFAIK it is not standardised. The internet draft for it seems to have expired.

That said, the CFRG document is in RG Last Call (if you look at the list archives), which indicates an intent to advance to publication (assuming no critical flaws are found in the review period).  The CFRG chairs are a bit understaffed at the moment (and looking for help), so it is unsurprising that documents are only progressing slowly.

**Comment by mbroz** [comment]
Argon2 RFC draft is in a new revision online: https://datatracker.ietf.org/doc/draft-irtf-cfrg-argon2/

**Comment by pubmikeb** [comment]
Argon2 is already mature enough, 5 years, it's a time to get it on board!

**Comment by johndoe31415** [comment]
I don't believe the question is regarding the maturity of Argon2 itself. The problem is with the lack of threading support of OpenSSL. Something that the currently open PR, #9444 adds as well. Arguably, the same goes for scrypt with p > 1 in the current implementation.

Anything that can be run in parallel (i.e., non-dependent computation) can trivially be sequentialized as well. However, it makes no sense because then the attacker has the full advantage and the honest implementation the disadvantage. Maybe counterintuitively, KDFs that run deliberately slow using an oftentimes parametizable work factor need to have *fast* implementations so that you're hurting your own performance but also that of the potential adversary equally much. You don't want to be the only one with a slow implementation while the attacker can make full use of parallel implementations.

**Comment by mattcaswell** [comment]
This is currently being progressed in #12255 and #12256.

**Comment by indo-dev-0** [comment]
>I don't believe the question is regarding the maturity of Argon2 itself. The problem is with the lack of threading support of OpenSSL. Something that the currently open PR, #9444 adds as well. Arguably, the same goes for scrypt with p > 1 in the current implementation.

Ignoring compatibility, you wouldn't want to remove Scrypt because it's *vastly* better than PBDFK2.  However, Argon2 has better sidechannel resistance and is more unfriendly to GPUs, FPGAs, and ASICs than Scrypt [[§1](https://password-hashing.net/argon2-specs.pdf)]....  

> Maybe counterintuitively, KDFs that run deliberately slow using an oftentimes parametizable work factor need to have fast implementations so that you're hurting your own performance but also that of the potential adversary equally much. You don't want to be the only one with a slow implementation while the attacker can make full use of parallel implementations.

I think the situation is more nuanced than that.  I've been running into difficulty [getting optimized versions of Argon2](https://github.com/hashcat/hashcat/issues/1966#issuecomment-477580890) because it's not popular enough. But I can't even get *password managers* to switch away from PBDFK2 :scream:  because [Argon2 is not widely available](https://community.bitwarden.com/t/switch-to-argon2/350/24) in their favorite crypto library (hence all the external bikeshedders).

But I also agree that `p > 1` on a serial implementation is a footgun :thinking:.  What if you limited support to Argon2id with a `p = 1`?  Then it _any_ usage of Argon2 would be strictly more secure than the alternatives and discussion/implementation of threading can be deferred.

**Comment by catleeball** [comment]
Hello! As an onlooker from outside who is excited about Argon2 KDF support, what action is currently needed on PRs #12255 and #12256 ? Are we just waiting for the 3.0 stable release before considering merging these PRs, or is there more work to be done?

Big shoutout to @ckalina for implementing the two PRs. :smiley: :tada:

**Comment by mattcaswell** [comment]
> Are we just waiting for the 3.0 stable release before considering merging these PRs, or is there more work to be done?

Basically everything is on hold until after 3.0 stable. So those PRs will have to target 3.1.

**Comment by catleeball** [comment]
Understood, thanks for the update @mattcaswell !

**Comment by mbroz** [comment]
Just FYI - released RFC for Argon2 is here: https://www.rfc-editor.org/rfc/rfc9106.txt

**Comment by odiferousmint** [comment]
We really need `Ardon2id`!

**Comment by mbroz** [comment]
Argon2id is part of the proposed code (and I successfully tested cryptsetup with patched OpenSSL 3.0.0).

The problem is that @ckalina has apparently no time to finish this to the state when merge request is acceptable ( #12255 and #12256).

**Comment by mbroz** [comment]
Both MRs (threads support #12255 and Argon2 #12256) are now rebased, I hope it is not going to stay there forever again :-) Please review it or comment there.

**Comment by t8m** [comment]
Oh yeah, we have Argon2 support in OpenSSL 3.2.0. Closing this!
